content/2_Bronze/Ad_Hoc.mdx

```diff
@@ -35,6 +36,186 @@ approaching problems that appear to be ad hoc.
 In the end, the best way to get better at ad hoc problems (or any type of
 problem) is to do a lot of them.
 
+## Example - Milking Order
+
+<FocusProblem problem="tutorial" />
+
+Don't be afraid to give up on some approach if you aren't making progress!
+
+<Spoiler title="Hint 1">
+Let's try placing cow $1$ at a specific position.
+
+How might we go about checking to see if we'll end up with a valid ordering?
+</Spoiler>
+
+
+### Solution
+
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_milkorder_bronze_open18.html)
+
+What if we tried placing cow $1$ at every possible position?
+
+Then, we'll have some hierarchy we have to fit in and some free cows which can go anywhere. Let's just handle the hierarchy, since we can fit in the free cows at the end.
+
+As we sweep through the hierarchy, we'll also store a pointer that indicates our current position. Greedily, we should try to place these cows as early as possible to make sure that we have room to fit in all of them. As we go through the list, we have to make sure that this pointer never outruns some previous cow in our hierarchy.
+
+This check takes $\mathcal{O}(N + M + K)$, which brings our total time complexity to $\mathcal{O}(N(N + M + K))$.
+
+<LanguageSection>
+<PySection>
+
+```py
+import sys
+
+sys.stdin = open("milkorder.in", "r")
+sys.stdout = open("milkorder.out", "w")
+
+n, m, k = map(int, input().split())
+
+hierarchy = [i - 1 for i in list(map(int, input().split()))]
+order = [-1] * n
+
+for i in range(k):
+	cow, pos = map(int, input().split())
+
+	order[pos - 1] = cow - 1
+
+	# already fixed, nothing we can do
+	if cow == 1:
+		print(pos)
+		exit(0)
+
+
+def check():
+	new_order = order.copy()
+
+	cow_to_pos = [-1] * n
+	for i in range(n):
+		if order[i] != -1:
+			cow_to_pos[order[i]] = i
+
+	h_idx = 0
+	i = 0
+	while i < n and h_idx < m:
+		# we know the next cow has to be in front of it
+		if cow_to_pos[hierarchy[h_idx]] != -1:
+			if i > cow_to_pos[hierarchy[h_idx]]:
+				return False
+
+			i = cow_to_pos[hierarchy[h_idx]]
+			h_idx += 1
+		else:
+			while i < n and new_order[i] != -1:
+				i += 1
+
+			# run out of places
+			if i == n:
+				return False
+
+			new_order[i] = hierarchy[h_idx]
+			cow_to_pos[hierarchy[h_idx]] = i
+			h_idx += 1
+
+		i += 1
+
+	return True
+
+
+for i in range(n):
+	# if already fixed, skip
+	if order[i] == -1:
+		# try placing cow 1 @ position i
+		order[i] = 0
+
+		if check():
+			print(i + 1)
+			break
+
+		order[i] = -1
+```
+</PySection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int n, m, k;
+
+bool check(vector<int> order, vector<int> &hierarchy) {
+	vector<int> cow_to_pos(n, -1);
+
+	for (int i = 0; i < n; i++) {
```

Comment:
(optional) remove globals and replace n & m w/ order.size()/hierarchy.size()

Reply:
did that initially, but i think it's way too messy bc we use `n` and `m` a lot o.o